File: app/src/main/java/com/example/zyyschedule/fragment/LabelFragment.java
package com.example.zyyschedule.fragment;

import android.app.AlertDialog;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.zyyschedule.R;
import com.example.zyyschedule.adapter.ScheduleAdapter;
import com.example.zyyschedule.database.Schedule;
import com.example.zyyschedule.databinding.FinishScheduleFootBinding;
import com.example.zyyschedule.databinding.FragmentLabelBinding;
import com.example.zyyschedule.databinding.ScheduleFootBinding;
import com.example.zyyschedule.databinding.ScheduleListFinishHeadBinding;
import com.example.zyyschedule.databinding.ScheduleListHeadBinding;
import com.example.zyyschedule.viewmodel.ScheduleViewModel;

import java.util.List;

public class LabelFragment extends Fragment {
    private  int labelId;
    private FragmentLabelBinding binding;
    private ScheduleViewModel vm;
    private ScheduleAdapter scheduleAdapter;
    private ScheduleAdapter finishScheduleAdapter;
    private List<Schedule> Schedules;
    private List<Schedule> finishSchedules;
    private ScheduleListHeadBinding scheduleHeadBinding;
    private ScheduleFootBinding scheduleFootBinding;
    private ScheduleListFinishHeadBinding scheduleListFinishHeadBinding;
    private FinishScheduleFootBinding finishScheduleFootBinding;
    private AlertDialog.Builder builder;

    public LabelFragment(int labelId) {
        this.labelId = labelId;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding  = DataBindingUtil.inflate(inflater,R.layout.fragment_label,container,false);
        scheduleHeadBinding = DataBindingUtil.inflate(inflater, R.layout.schedule_list_head, container, false);
        scheduleFootBinding = DataBindingUtil.inflate(inflater, R.layout.schedule_foot, container, false);
        scheduleListFinishHeadBinding = DataBindingUtil.inflate(inflater, R.layout.schedule_list_finish_head, container, false);
        finishScheduleFootBinding = DataBindingUtil.inflate(inflater, R.layout.finish_schedule_foot, container, false);
        return binding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        vm = new ViewModelProvider(this).get(ScheduleViewModel.class);
        builder = new AlertDialog.Builder(getContext());
        LinearLayoutManager labelSchedule =  new LinearLayoutManager(getContext());
        labelSchedule.setOrientation(LinearLayoutManager.VERTICAL);
        LinearLayoutManager finishSchedule =  new LinearLayoutManager(getContext());
        finishSchedule.setOrientation(LinearLayoutManager.VERTICAL);
        binding.labelScheduleList.setLayoutManager(labelSchedule);
        binding.finishScheduleList.setLayoutManager(finishSchedule);
        scheduleAdapter = new ScheduleAdapter(R.layout.schedule_item_local);
        scheduleAdapter.setOwner(this);
        scheduleAdapter.setHeaderView(scheduleHeadBinding.getRoot());
        scheduleAdapter.addFooterView(scheduleFootBinding.getRoot());
        finishScheduleAdapter = new ScheduleAdapter(R.layout.schedule_item_local);
        finishScheduleAdapter.setOwner(this);
        finishScheduleAdapter.setHeaderView(scheduleListFinishHeadBinding.getRoot());
        finishScheduleAdapter.setFooterView(finishScheduleFootBinding.getRoot());
        binding.labelScheduleList.setAdapter(scheduleAdapter);
        binding.finishScheduleList.setAdapter(finishScheduleAdapter);
        scheduleHeadBinding.scheduleListHead.setText(R.string.local_schedule_head);
        scheduleHeadBinding.scheduleDeleteBack.setOnClickListener(v -> {
            scheduleHeadBinding.deleteSchedule.setVisibility(View.GONE);
            scheduleHeadBinding.scheduleListHead.setVisibility(View.VISIBLE);
            scheduleHeadBinding.scheduleDeleteBack.setVisibility(View.GONE);
            for (int i = 0; i < Schedules.size(); i++) {
                Schedules.get(i).setEditor(false);
            }
            for (int i = 0; i < finishSchedules.size(); i++) {
                finishSchedules.get(i).setEditor(false);
            }
            scheduleAdapter.notifyDataSetChanged();
            finishScheduleAdapter.notifyDataSetChanged();
        });
        scheduleListFinishHeadBinding.scheduleDeleteBack.setOnClickListener(v -> {
            scheduleListFinishHeadBinding.deleteSchedule.setVisibility(View.GONE);
            scheduleListFinishHeadBinding.scheduleListFinish.setVisibility(View.VISIBLE);
            scheduleListFinishHeadBinding.scheduleDeleteBack.setVisibility(View.GONE);
            for (int i = 0; i < Schedules.size(); i++) {
                Schedules.get(i).setEditor(false);
            }
            for (int i = 0; i < finishSchedules.size(); i++) {
                finishSchedules.get(i).setEditor(false);
            }
            scheduleAdapter.notifyDataSetChanged();
            finishScheduleAdapter.notifyDataSetChanged();
        });
        scheduleHeadBinding.deleteSchedule.setOnClickListener(v -> gotoDeleteDialog());
        scheduleListFinishHeadBinding.deleteSchedule.setOnClickListener(v -> gotoDeleteDialog());
        //完成和未完成日程item长按事件
        scheduleAdapter.setOnItemLongClickListener((adapter, view, position) -> {
            scheduleHeadBinding.deleteSchedule.setVisibility(View.VISIBLE);
            scheduleHeadBinding.scheduleListHead.setVisibility(View.GONE);
            scheduleHeadBinding.scheduleDeleteBack.setVisibility(View.VISIBLE);
            for (int i = 0; i < Schedules.size(); i++) {
                Schedules.get(i).setEditor(true);
            }
            for (int i = 0; i < finishSchedules.size(); i++) {
                finishSchedules.get(i).setEditor(true);
            }
            adapter.notifyDataSetChanged();
            finishScheduleAdapter.notifyDataSetChanged();
            return true;
        });
        finishScheduleAdapter.setOnItemLongClickListener((adapter, view, position) -> {
            scheduleHeadBinding.deleteSchedule.setVisibility(View.VISIBLE);
            scheduleHeadBinding.scheduleListHead.setVisibility(View.GONE);
            scheduleHeadBinding.scheduleDeleteBack.setVisibility(View.VISIBLE);
            if (scheduleAdapter.getData().size() == 0) {
                scheduleListFinishHeadBinding.deleteSchedule.setVisibility(View.VISIBLE);
                scheduleListFinishHeadBinding.scheduleListFinish.setVisibility(View.GONE);
                scheduleListFinishHeadBinding.scheduleDeleteBack.setVisibility(View.VISIBLE);
            }
            for (int i = 0; i < Schedules.size(); i++) {
                Schedules.get(i).setEditor(true);
            }
            for (int i = 0; i < finishSchedules.size(); i++) {
                finishSchedules.get(i).setEditor(true);
            }
            adapter.notifyDataSetChanged();
            scheduleAdapter.notifyDataSetChanged();
            return true;
        });
        UpdateScheduleList();
    }

    public void gotoDeleteDialog() {
        builder = new AlertDialog.Builder(getContext());
        builder.setMessage(R.string.delete_schedule_message)
                .setPositiveButton(R.string.dialog_button_ok, (dialog, which) -> {
                    for (int i = 0; i < Schedules.size(); i++) {
                        if (Schedules.get(i).isEditorChecked()) {
                            vm.deleteSchedule(Schedules.get(i));
                        }
                    }
                    for (int i = 0; i < finishSchedules.size(); i++) {
                        if (finishSchedules.get(i).isEditorChecked()) {
                            vm.deleteSchedule(finishSchedules.get(i));
                        }
                    }
                    dialog.dismiss();
                    UpdateScheduleList();
                    scheduleListFinishHeadBinding.deleteSchedule.setVisibility(View.GONE);
                    scheduleListFinishHeadBinding.scheduleListFinish.setVisibility(View.VISIBLE);
                    scheduleListFinishHeadBinding.scheduleDeleteBack.setVisibility(View.GONE);
                    scheduleHeadBinding.deleteSchedule.setVisibility(View.GONE);
                    scheduleHeadBinding.scheduleListHead.setVisibility(View.VISIBLE);
                    scheduleHeadBinding.scheduleDeleteBack.setVisibility(View.GONE);
                })
                .setNeutralButton(R.string.dialog_button_cancel, (dialog, which) -> dialog.dismiss());
        builder.create().show();
    }

    private void UpdateScheduleList() {
        vm.getUFScheduleOfLabel(labelId).observe(getViewLifecycleOwner(), schedules -> {
            for (int i = 0; i < schedules.size(); i++) {
                schedules.get(i).setChecked(false);
            }
            if (schedules.size() == 0) {
                scheduleHeadBinding.getRoot().setVisibility(View.GONE);
                scheduleFootBinding.getRoot().setVisibility(View.GONE);
            } else {
                scheduleHeadBinding.getRoot().setVisibility(View.VISIBLE);
                scheduleFootBinding.getRoot().setVisibility(View.VISIBLE);
            }
            scheduleAdapter.setList(schedules);
            Schedules = scheduleAdapter.getData();
        });
        vm.getFScheduleOfLabel(labelId).observe(getViewLifecycleOwner(), schedules -> {
            for (int i = 0; i < schedules.size(); i++) {
                schedules.get(i).setChecked(true);
            }
            if (schedules.size() == 0) {
                scheduleListFinishHeadBinding.getRoot().setVisibility(View.GONE);
                finishScheduleFootBinding.getRoot().setVisibility(View.GONE);
            } else {
                scheduleListFinishHeadBinding.getRoot().setVisibility(View.VISIBLE);
                finishScheduleFootBinding.getRoot().setVisibility(View.VISIBLE);
            }
            finishScheduleAdapter.setList(schedules);
            finishSchedules = finishScheduleAdapter.getData();
        });
    }


}

File: app/src/main/java/com/example/zyyschedule/fragment/LabelFragment.kt
package com.example.zyyschedule.fragment

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.DrawableCompat
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.LinearLayoutManager
import com.chad.library.adapter.base.BaseQuickAdapter
import com.example.zyyschedule.R
import com.example.zyyschedule.adapter.ScheduleAdapter
import com.example.zyyschedule.database.Schedule
import com.example.zyyschedule.databinding.*
import com.example.zyyschedule.viewmodel.ScheduleViewModel
import com.jeremyliao.liveeventbus.LiveEventBus

@SuppressLint("NotifyDataSetChanged")
class LabelFragment(labelId: Int) : Fragment(), View.OnClickListener {
    private val mLabelId = labelId
    private lateinit var binding: FragmentLabelBinding
    private val vm: ScheduleViewModel by viewModels()
    private lateinit var scheduleAdapter: ScheduleAdapter
    private lateinit var finishScheduleAdapter: ScheduleAdapter
    private lateinit var mSchedules: List<Schedule>
    private lateinit var mFinishSchedules: List<Schedule>
    private lateinit var scheduleHeadBinding: ScheduleListHeadBinding
    private lateinit var scheduleFootBinding: ScheduleFootBinding
    private lateinit var scheduleListFinishHeadBinding: ScheduleListFinishHeadBinding
    private lateinit var finishScheduleFootBinding: FinishScheduleFootBinding
    private lateinit var builder: AlertDialog.Builder

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_label, container, false)
        scheduleHeadBinding =
            DataBindingUtil.inflate(inflater, R.layout.schedule_list_head, container, false)
        scheduleFootBinding =
            DataBindingUtil.inflate(inflater, R.layout.schedule_foot, container, false)
        scheduleListFinishHeadBinding =
            DataBindingUtil.inflate(inflater, R.layout.schedule_list_finish_head, container, false)
        finishScheduleFootBinding =
            DataBindingUtil.inflate(inflater, R.layout.finish_schedule_foot, container, false)
        return binding.root
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.deleteButton.setOnClickListener(this)
        builder = AlertDialog.Builder(context)
        val labelSchedule = LinearLayoutManager(context)
        labelSchedule.orientation = LinearLayoutManager.VERTICAL
        val finishSchedule = LinearLayoutManager(context)
        finishSchedule.orientation = LinearLayoutManager.VERTICAL
        binding.labelScheduleList.layoutManager = labelSchedule
        binding.finishScheduleList.layoutManager = finishSchedule
        scheduleAdapter = ScheduleAdapter(R.layout.schedule_item_local)
        scheduleAdapter.setOwner(this)
        scheduleAdapter.setHeaderView(scheduleHeadBinding.root)
        scheduleAdapter.addFooterView(scheduleFootBinding.root)
        finishScheduleAdapter = ScheduleAdapter(R.layout.schedule_item_local)
        finishScheduleAdapter.setOwner(this)
        finishScheduleAdapter.setHeaderView(scheduleListFinishHeadBinding.root)
        finishScheduleAdapter.setFooterView(finishScheduleFootBinding.root)
        binding.labelScheduleList.adapter = scheduleAdapter
        binding.finishScheduleList.adapter = finishScheduleAdapter
        scheduleHeadBinding.scheduleListHead.setText(R.string.local_schedule_head)
        LiveEventBus.get("ScheduleF_SomeF", String::class.java)
            .observe(viewLifecycleOwner, {
                when (it) {
                    "adapterComeBack" -> {
                        for (i in mSchedules.indices) {
                            mSchedules[i].isEditor = false
                        }
                        for (i in mFinishSchedules.indices) {
                            mFinishSchedules[i].isEditor = false
                        }
                        scheduleAdapter.notifyDataSetChanged()
                        finishScheduleAdapter.notifyDataSetChanged()
                        binding.editorLayout.visibility = View.GONE
                    }
                }
            })
        //编辑模式下选中的监听
        scheduleAdapter.pitchOnNumber.observe(viewLifecycleOwner, {
            LiveEventBus
                .get("pitchOnNumber", Int::class.java)
                .post(it)
            if (it > 0) {
                enabledTrue()
            } else {
                enabledFalse()
            }
        })
        finishScheduleAdapter.pitchOnNumber.observe(viewLifecycleOwner, {
            LiveEventBus
                .get("pitchOnNumber", Int::class.java)
                .post(it)
            if (it > 0) {
                enabledTrue()
            } else {
                enabledFalse()
            }
        })
        //完成和未完成日程item长按事件
        scheduleAdapter.setOnItemLongClickListener { adapter: BaseQuickAdapter<*, *>, _: View?, _: Int ->
            binding.editorLayout.visibility = View.VISIBLE
            LiveEventBus
                .get("SomeF_ScheduleF", String::class.java)
                .post("gone_titleBar")
            LiveEventBus
                .get("SomeF_MainA", String::class.java)
                .post("gone_navigation")
            for (i in mSchedules.indices) {
                mSchedules[i].isEditor = true
            }
            for (i in mFinishSchedules.indices) {
                mFinishSchedules[i].isEditor = true
            }
            adapter.notifyDataSetChanged()
            finishScheduleAdapter.notifyDataSetChanged()
            true
        }
        finishScheduleAdapter.setOnItemLongClickListener { adapter: BaseQuickAdapter<*, *>, _: View?, _: Int ->
            binding.editorLayout.visibility = View.VISIBLE
            LiveEventBus
                .get("SomeF_ScheduleF", String::class.java)
                .post("gone_titleBar")
            LiveEventBus
                .get("SomeF_MainA", String::class.java)
                .post("gone_navigation")
            for (i in mSchedules.indices) {
                mSchedules[i].isEditor = true
            }
            for (i in mFinishSchedules.indices) {
                mFinishSchedules[i].isEditor = true
            }
            adapter.notifyDataSetChanged()
            scheduleAdapter.notifyDataSetChanged()
            true
        }
        updateScheduleList()
    }

    override fun onClick(v: View?) {
        v?.let {
            when (it.id) {
                R.id.delete_button -> gotoDeleteDialog()
            }
        }
    }

    private fun gotoDeleteDialog() {
        builder = AlertDialog.Builder(context)
        builder.setMessage(R.string.delete_schedule_message)
            .setPositiveButton(R.string.dialog_button_ok) { dialog, _ ->
                for (i in mSchedules.indices) {
                    if (mSchedules[i].isEditorChecked) {
                        vm.deleteSchedule(mSchedules[i])
                    }
                }
                for (i in mFinishSchedules.indices) {
                    if (mFinishSchedules[i].isEditorChecked) {
                        vm.deleteSchedule(mFinishSchedules[i])
                    }
                }
                dialog.dismiss()
                updateScheduleList()
                binding.editorLayout.visibility = View.GONE
                LiveEventBus.get("SomeF_ScheduleF", String::class.java)
                    .post("visibility_titleBar")
            }
            .setNeutralButton(R.string.dialog_button_cancel) { dialog, _ -> dialog.dismiss() }
        builder.create().show()
    }

    private fun updateScheduleList() {
        vm.getUFScheduleOfLabel(mLabelId).observe(viewLifecycleOwner, { schedules: List<Schedule> ->
            for (i in schedules.indices) {
                schedules[i].isChecked = false
            }
            if (schedules.isEmpty()) {
                scheduleHeadBinding.root.visibility = View.GONE
                scheduleFootBinding.root.visibility = View.GONE
            } else {
                scheduleHeadBinding.root.visibility = View.VISIBLE
                scheduleFootBinding.root.visibility = View.VISIBLE
            }
            scheduleAdapter.setList(schedules)
            mSchedules = scheduleAdapter.data
        })
        vm.getFScheduleOfLabel(mLabelId).observe(viewLifecycleOwner, { schedules: List<Schedule> ->
            for (i in schedules.indices) {
                schedules[i].isChecked = true
            }
            if (schedules.isEmpty()) {
                scheduleListFinishHeadBinding.root.visibility = View.GONE
                finishScheduleFootBinding.root.visibility = View.GONE
            } else {
                scheduleListFinishHeadBinding.root.visibility = View.VISIBLE
                finishScheduleFootBinding.root.visibility = View.VISIBLE
            }
            finishScheduleAdapter.setList(schedules)
            mFinishSchedules = finishScheduleAdapter.data
            finishScheduleAdapter.otherDate = scheduleAdapter.data
            scheduleAdapter.otherDate = finishScheduleAdapter.data
        })
    }

    //控制编辑栏按钮可以点击
    private fun enabledTrue() {
        binding.deleteButton.isClickable = true
        binding.moreButton.isClickable = true
        binding.labelButton.isClickable = true
        binding.timeButton.isClickable = true
        ContextCompat.getDrawable(requireContext(), R.drawable.ic_baseline_delete_outline_24)?.let {
            DrawableCompat.setTint(it, Color.BLACK)
            binding.deleteButton.setImageDrawable(it)
        }
        ContextCompat.getDrawable(requireContext(), R.drawable.ic_baseline_more_24)?.let {
            DrawableCompat.setTint(it, Color.BLACK)
            binding.moreButton.setImageDrawable(it)
        }
        ContextCompat.getDrawable(requireContext(), R.drawable.ic_schedule_24)?.let {
            DrawableCompat.setTint(it, Color.BLACK)
            binding.labelButton.setImageDrawable(it)
        }
        ContextCompat.getDrawable(requireContext(), R.drawable.ic_calendar_toolbar)?.let {
            DrawableCompat.setTint(it, Color.BLACK)
            binding.timeButton.setImageDrawable(it)
        }
    }

    //控制编辑栏按钮不可点击
    private fun enabledFalse() {
        binding.deleteButton.isClickable = false
        binding.moreButton.isClickable = false
        binding.labelButton.isClickable = false
        binding.timeButton.isClickable = false
        ContextCompat.getDrawable(requireContext(), R.drawable.ic_baseline_delete_outline_24)?.let {
            DrawableCompat.setTint(it, Color.GRAY)
            binding.deleteButton.setImageDrawable(it)
        }
        ContextCompat.getDrawable(requireContext(), R.drawable.ic_baseline_more_24)?.let {
            DrawableCompat.setTint(it, Color.GRAY)
            binding.moreButton.setImageDrawable(it)
        }
        ContextCompat.getDrawable(requireContext(), R.drawable.ic_schedule_24)?.let {
            DrawableCompat.setTint(it, Color.GRAY)
            binding.labelButton.setImageDrawable(it)
        }
        ContextCompat.getDrawable(requireContext(), R.drawable.ic_calendar_toolbar)?.let {
            DrawableCompat.setTint(it, Color.GRAY)
            binding.timeButton.setImageDrawable(it)
        }
    }


}